### PR TITLE
Logic Conditions framework

### PR DESCRIPTION
**Pull request by DzikuVx**

### CODE REVIEW

**Comment by rb1205** [comment]
Is this intended for convertiplanes / VTOL flight?

**Comment by DzikuVx** [comment]
@rb1205 VTOL is only one reason for it. This will solve the problem of swapped control axises after conversion from vertical to horizontal flight, but only this one

**Comment by hali9** [comment]
Do you plan to do the same for mmix?
It would be necessary to do it for mmix for VTOL.

**Comment by DzikuVx** [comment]
@hali9 yes

File: src/main/fc/cli.c [hunk 1]
```diff
@@ -1501,30 +1501,44 @@ static void printServoMix(uint8_t dumpMask, const servoMixer_t *customServoMixer
...
+        #endif
         );
     }
 }
 
 static void cliServoMix(char *cmdline)
 {
     char * saveptr;
-    int args[8], check = 0;
+    int args[10], check = 0;
```
Comment (shellixyz) [inline src/main/fc/cli.c]: Why 10 ? Looks like 6 would be enough.

File: src/main/fc/cli.c [hunk 2]
```diff
@@ -1546,22 +1560,134 @@ static void cliServoMix(char *cmdline)
...
+            logic.operandB.type,
+            logic.operandB.value,
+            logic.flags
+        );
+    }
+}
+
+static void cliLogic(char *cmdline) {
+    char * saveptr;
+    int args[10], check = 0;
```
Comment (shellixyz) [inline src/main/fc/cli.c]: Why 10 ? Looks like 8 would be enough

File: src/main/fc/cli.c [hunk 3]
```diff
@@ -1546,22 +1560,134 @@ static void cliServoMix(char *cmdline)
...
+        ) {
+            logicConditionsMutable(i)->enabled = args[ENABLED];
+            logicConditionsMutable(i)->operation = args[OPERATION];
+            logicConditionsMutable(i)->operandA.type = args[OPERAND_A_TYPE];
+            logicConditionsMutable(i)->operandA.value = args[OPERAND_A_VALUE];
+            logicConditionsMutable(i)->operandB.type = args[OPERAND_B_TYPE];
+            logicConditionsMutable(i)->operandB.value = args[OPERAND_B_VALUE];
+            logicConditionsMutable(i)->flags = args[FLAGS];
+
+            cliLogic("");
```
Comment (shellixyz) [inline src/main/fc/cli.c:1681]: Is that intended or you put this only for debugging ?

File: src/main/fc/cli.c [hunk 4]
```diff
@@ -1546,22 +1560,134 @@ static void cliServoMix(char *cmdline)
...
+    char * saveptr;
+    int args[10], check = 0;
+    uint8_t len = strlen(cmdline);
+
+    if (len == 0) {
+        printLogic(DUMP_MASTER, logicConditions(0), NULL);
+    } else if (sl_strncasecmp(cmdline, "reset", 5) == 0) {
+        pgResetCopy(logicConditionsMutable(0), PG_SERVO_MIXER);
+    } else {
+        //TODO implement
```
Comment (shellixyz) [inline src/main/fc/cli.c]: Leftover ?

File: src/main/fc/fc_msp.c [hunk 5]
```diff
@@ -1823,7 +1849,38 @@ static mspResult_e mspFcProcessInCommand(uint16_t cmdMSP, sbuf_t *src)
         } else
             return MSP_RESULT_ERROR;
         break;
-
+    
+    case MSP2_INAV_SET_SERVO_MIXER:
+        sbufReadU8Safe(&tmp_u8, src);
+        if ((dataSize == 12) && (tmp_u8 < MAX_SERVO_RULES)) {
```
Comment (shellixyz) [inline src/main/fc/fc_msp.c]: 12 ? You're reading only 6

File: src/main/fc/fc_msp.c [hunk 6]
```diff
@@ -1823,7 +1849,38 @@ static mspResult_e mspFcProcessInCommand(uint16_t cmdMSP, sbuf_t *src)
...
+            customServoMixersMutable(tmp_u8)->targetChannel = sbufReadU8(src);
+            customServoMixersMutable(tmp_u8)->inputSource = sbufReadU8(src);
+            customServoMixersMutable(tmp_u8)->rate = sbufReadU8(src);
```
Comment (shellixyz) [inline src/main/fc/fc_msp.c]: Rate should be int16 now

File: src/main/fc/fc_msp.c [hunk 7]
```diff
@@ -463,6 +463,32 @@ static bool mspFcProcessOutCommand(uint16_t cmdMSP, sbuf_t *dst, mspPostProcessF
             sbufWriteU8(dst, 0);
         }
         break;
+    case MSP2_INAV_SERVO_MIXER:
+        for (int i = 0; i < MAX_SERVO_RULES; i++) {
+            sbufWriteU8(dst, customServoMixers(i)->targetChannel);
+            sbufWriteU8(dst, customServoMixers(i)->inputSource);
+            sbufWriteU8(dst, customServoMixers(i)->rate);
```
Comment (shellixyz) [inline src/main/fc/fc_msp.c]: Please replace `sbufWriteU8` with `sbufWriteU16`. The servo mixer rate is now int16

**Review by shellixyz — CHANGES_REQUESTED** [review]
Otherwise LGTM

**Review by shellixyz — APPROVED** [review]

**Comment by stronnag** [comment]
small fyi,
```
%% logic_condition.c 
./src/main/common/logic_condition.c: In function 'logicConditionUpdateTask':
./src/main/common/logic_condition.c:237:40: warning: unused parameter 'currentTimeUs' [-Wunused-parameter]
 void logicConditionUpdateTask(timeUs_t currentTimeUs) {
                               ~~~~~~~~~^~~~~~~~~~~~~
```

**Comment by stronnag** [comment]
~~I currently (505f44ad8) have to set~~

~~logic 0 1 0 0 0 0 0 0~~

~~(whatever that does :) to have working servos. Is this expected ?~~
~~Shouldn't the default be "work as legacy".~~
 .. `smix .... s/0/-1/` , aha.

**Comment by DzikuVx** [comment]
yes, `-1` with `smix`. I will check defaults tho

**Comment by stronnag** [comment]
On March 19, 2019 7:45:04 AM UTC, "Paweł Spychalski" <notifications@github.com> wrote:
>yes, `-1` with `smix`. I will check defaults tho
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/iNavFlight/inav/pull/4144#issuecomment-474231953

Thanks. As I'd just restored a 0,0 diff, user error not checking what had changed. Probably saved me from flying a light, floaty wing in far too much wind.

**Pull request merged**